### ISSUE TITLE
Breaks in multiple ways with rust binaries with debug symbols left in

### ISSUE THREAD

**Issue by winneratwin** [korcankaraokcu/PINCE#217]
this is a combination of the gdb python bindings being bad and the fact that PINCE doesn't have support for languages with no void types

`GDBCommandExtensions.py line 23` breaks because the gdb python bindings can't convert `&str` types (which environment variables are cast to in rust) to strings you can fix it by replacing `.string()` with `.format_string().strip('"')`
same for `ScriptUtils.py line 22 and 23`

if you fix those lines you get a new error at `ScriptUtils.py line 33` since it tries to get the void type but fails since rust has no void/null type and even if you comment it out it fails at a more systematic level where `RegisterLabel.py line 29` is erroring with this error `can only concatenate str (not "NoneType") to str` which i am assuming is it trying to do something with the void type at which point i gave up trying to fix PINCE and decided to make an issue

**Comment by korcankaraokcu** [comment]
> languages with no void types

Completely slipped my attention, thanks for the bug report. I'll look into this soon

**Comment by brkzlr** [comment]
I don't think this has to do with whether a language supports void or not as I tested it on a Go binary, where Go also doesn't have a void type, and no issues occur.

I think this has more to do with gdb's python binding that has poor support for Rust more than the void-ness of a language.

**Comment by korcankaraokcu** [comment]
Tested with a simple, infinitely looping rust binary with debug info. `.format_string().strip('"')` doesn't fix anything, GDB can't read the path var at all, printing the variable gives `<error reading variable: Cannot access memory at address 0x8>`

**Comment by korcankaraokcu** [comment]
I've solved the void pointer issue with changing this code In ScriptUtils

`void_ptr = gdb.lookup_type("void").pointer()` to `void_ptr = gdb.lookup_type("u8").pointer()`

It works if we use u8 instead of void. But again, it fails in many places that uses C-like expressions, such as breakpoints and watchpoints. Maybe it's about the time to refactor the older codebase to use more generalized expressions. However, this might take a while

**Comment by korcankaraokcu** [comment]
Here's something interesting, none of these issues appear if the target is compiled in release mode

**Comment by korcankaraokcu** [comment]
Considering all other issues, such as not being able to read string variables and all. I think the best solution would be to make GDB act like it's a release binary instead of trying to fix every single edge case out there

**Comment by brkzlr** [comment]
This needs a re-test on recent versions

**Comment by brkzlr** [comment]
Fixed in 530d8e3a22c074024f126795f29b13d16eb167ae.

Indeed as Issue author mentions, gdb's python bindings are pretty bad and full of hacks that only apply to certain languages and not to others. This issue is caused by both the python bindings and PINCE's assumed void type.

Basically the main issue was **`gdb.Value.string()`** calling **`c_get_string`** (`gdb/c-lang.c:242`), which only accepts **`TYPE_CODE_ARRAY`** (`char[N]`) and **`TYPE_CODE_PTR`** (`char*`).

Anything else (struct, slice, etc...) raises `Trying to read string with inappropriate type {}`.

### Why only debug Rust?
- Debug: DWARF carries the language tag -> GDB's Rust module infers "..." as `&str` which is a {data, len} struct, which results in **`TYPE_CODE_STRUCT`** which results in  .string() rejecting it.
  - This also happens for any `&str` field accessed via `examine_expression`.
- Release: strips all DWARF info so there's no language tag -> GDB falls back to C-like literal inference which results in "..." becoming `char[N]` causing .string() to work.

### Why Go seemed fine at first
Go's `go-lang.c` GDB language module does two things Rust's `rust-lang.c` doesn't:
- Infers string literals as `char[]` (**`TYPE_CODE_ARRAY`**), not as the string struct so the convenience-var read path works.
- It also carries the C basic-type names in DWARF so `lookup_type("void")` works.

So the void type lookup and the convenience var read/set don't raise on Go, but there still was one issue that I didn't catch in my original post.

As an example, running `ptype player.Name` shows `struct string { uint8 *str; int len; }` and `.string()` will raise the same `Trying to read string with inappropriate type {}` error that fires on Rust's `&str`.

### Is it just these languages?
***No, this can happen on any other language that does not map the string literals to either a C `char[]` or `char*`*** or if the debug DWARF information omits C basic type names.

I'd suspect this would also happen on Zig (which is similar to how Rust represents strings), Swift and others so it wasn't just Rust being fixed by this.
